pkg/ebpf/filter.go

```diff
@@ -0,0 +1,429 @@
+package ebpf
+
+import (
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/aquasecurity/tracee/pkg/events"
+	"github.com/aquasecurity/tracee/pkg/filters"
+	"github.com/aquasecurity/tracee/types/protocol"
+)
+
+const (
+	UIDFilterMap         = "uid_filter"
+	PIDFilterMap         = "pid_filter"
+	MntNSFilterMap       = "mnt_ns_filter"
+	PidNSFilterMap       = "pid_ns_filter"
+	UTSFilterMap         = "uts_ns_filter"
+	CommFilterMap        = "comm_filter"
+	ProcessTreeFilterMap = "process_tree_map"
+	CgroupIdFilterMap    = "cgroup_id_filter"
+	ContIdFilter         = "cont_id_filter"
+)
+
+type Filter struct {
+	EventsToTrace     []events.ID
+	UIDFilter         *filters.BPFUIntFilter
+	PIDFilter         *filters.BPFUIntFilter
+	NewPidFilter      *filters.BoolFilter
+	MntNSFilter       *filters.BPFUIntFilter
+	PidNSFilter       *filters.BPFUIntFilter
+	UTSFilter         *filters.BPFStringFilter
+	CommFilter        *filters.BPFStringFilter
+	ContFilter        *filters.BoolFilter
+	NewContFilter     *filters.BoolFilter
+	ContIDFilter      *filters.ContainerFilter
+	RetFilter         *filters.RetFilter
+	ArgFilter         *filters.ArgFilter
+	ProcessTreeFilter *filters.ProcessTreeFilter
+	Follow            bool
+	NetFilter         *NetIfaces
+}
+
+func IsValidFilterField(field string) bool {
+	if strings.Contains(field, ".") {
+		parts := strings.Split(field, ".")
+		first := parts[0]
+		if _, ok := events.Definitions.GetID(first); !ok {
+			return false
+		}
+		return true
+	}
+
+	switch field {
+	case "comm", "processName", "container", "container/new", "event", "set", "pid", "pid/new", "uts", "uid", "follow", "mntns", "pidns", "net":
+		return true
+	}
+
+	return false
+}
+
+func buildFilterError(filterName string, err error) error {
+	return fmt.Errorf("failed to build %s filter: %s", filterName, err)
+}
+
+func ParseProtocolFilters(filterRequests []protocol.Filter) (Filter, error) {
+	// aggregate protocol.Filter by "topics" (inner filters in tracee)
+	const (
+		retval      = "retval"
+		args        = "args"
+		event       = "events"
+		comm        = "comm"
+		container   = "container"
+		newcont     = "newcont"
+		containerid = "containerid"
+		sets        = "set"
+		pid         = "pid"
+		newpid      = "newpid"
+		uts         = "uts"
+		uid         = "uid"
+		follow      = "follow"
+		mntns       = "mntns"
+		pidns       = "pidns"
+		net         = "net"
+	)
+	filterMap := map[string][]protocol.Filter{
+		retval:    {},
+		args:      {},
+		event:     {},
+		comm:      {},
+		container: {},
+		newcont:   {},
+		sets:      {},
+		pid:       {},
+		newpid:    {},
+		uts:       {},
+		follow:    {},
+		mntns:     {},
+		pidns:     {},
+		net:       {},
+	}
+
+	for _, filterReq := range filterRequests {
+		field := filterReq.Field
+		if strings.HasSuffix(field, ".retval") {
+			filterMap[retval] = append(filterMap[retval], filterReq)
+			continue
+		}
+
+		if strings.Contains(field, ".") {
+			filterMap[args] = append(filterMap[retval], filterReq)
```

Comment:
Possibly a missing test for these cases, will fix and add

Reply:
Tests didn't catch it because we test only adding one argfilter, if so the retval map is empty so the test passes since the input is fine. Case missing is adding both a retval and an arg filter